modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.future;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Maintains a collection of in-flight {@link CompletableFuture}s (i.e. futures that are not yet completed) to later
+ * have a possibility to process them somehow (for example, cancel them all if we know for sure that they will
+ * never be completed).
+ */
+public class InFlightFutures implements Iterable<CompletableFuture<?>> {
+    private final Set<CompletableFuture<?>> inFlightFutures = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    /**
+     * Registers a future in the in-flight futures collection. When it completes (either normally or exceptionally),
+     * it will be removed from the collection.
+     *
+     * @param future the future to register
+     */
+    public void registerFuture(CompletableFuture<?> future) {
+        future.whenComplete((result, ex) -> inFlightFutures.remove(future));
```

Comment:
And what about pipelining? The ordering of independent completion stages isn't guaranteed, so it's possible to remove the future while one of it's stages is still in progress. E.g.
```
CompletableFuture<R> future = startLocalExecution(jobClass, args);
inFlightFutures.registerFuture(future);
...
future.thenAcceptAsync()
```

Reply:
This PR is about making sure that a future does not get stuck due to our thread pool being stopped. It looks like the consequtive stages should be responsibility of those who add them.

There is also another problem: consequtive stages might be computationally heavy, and they can potentially block our 'job execution thread pool'. This needs further investigation, I've created an issue: https://issues.apache.org/jira/browse/IGNITE-16747